### ISSUE TITLE
Is Pipeline Parallelism same as Parameter partitioning?

### ISSUE THREAD

**Issue by ghosthamlet** [microsoft/DeepSpeed#505]
Thanks for the amazing DeepSpeed! I can train/finetune GPT2 1.5B on a single 2080Ti, with a few train code change.

Your new feature Pipeline Parallelism is very promising, is it the Parameter Partitioning in ZeRO paper?
If they are not the same, what are the Pros and Cons of Pipeline Parallelism and Parameter Partitioning?

**Comment by ShadenSmith** [comment]
Thanks for your interest in DeepSpeed!

Pipeline parallelism is not the same thing as ZeRO's parameter partitioning, though they accomplish similar goals. Pipeline parallelism also partitions a model (and gradients, and optimizer states) among GPUs. The tradeoff is that your model needs to have sufficient depth to be partitioned (i.e., it needs to have enough layers), and `forward()` must follow a specific format.

Pipeline parallelism is a big topic with several algorithms such as [PipeDream](https://arxiv.org/abs/1806.03377) and [GPipe](https://papers.nips.cc/paper/2019/file/093f65e080a295f8076b1c5722a46aa2-Paper.pdf). DeepSpeed uses a different algorithm that trades off a batch size for the same convergence properties as traditional data parallelism. Our [pipeline tutorial](https://www.deepspeed.ai/tutorials/pipeline/) covers the specific algorithm that we use in DeepSpeed today. A schedule like PipeDream should be possible as well if the schedule was implemented.

ZeRO does not include parameter partitioning yet, but once released they can even be composed. Pipeline parallelism works with ZeRO stage 1 and 2 today (optimizer and gradient partitioning). ZeRO-3 (parameter partitioning) will be more simple to use because it does not require model modification, but introduces additional communication. Pipeline parallelism on the other hand may require some model modifications but typically improves communication overheads. We've seen some large speedups with pipeline parallelism on low-bandwidth networks.

**Comment by ghosthamlet** [comment]
Thanks for the detailed answer, I got it now.

**Comment by g-karthik** [comment]
@ShadenSmith is there an ETA by which y'all are planning to release Stage 3? I'd love to beta-test it if y'all have it on some other branch!
